[ja] Update Error page (#22163)
mdn/translated-content: files/ja/web/javascript/reference/global_objects/error/index.md

### 一般的なエラーを発生させる

通常、{{jsxref("Statements/throw", "throw")}} キーワードを使い意図的にエラーを発生させて `Error` オブジェクトを生成します。 {{jsxref("Statements/try...catch", "try...catch")}} 構文を使用してエラーを処理してください:

```js
try {
  throw new Error("Whoops!");
} catch (e) {
  console.error(`${e.name}: ${e.message}`);
}
```

### 特定のエラーを処理する

{{jsxref("Operators/instanceof", "instanceof")}} でエラー型をテストすることにより、特定のエラー型だけを選んで処理できます:

```js
try {
  foo.bar();
} catch (e) {
  if (e instanceof EvalError) {
    console.error(`${e.name}: ${e.message}`);
  } else if (e instanceof RangeError) {
    console.error(`${e.name}: ${e.message}`);
  }
  // など
  else {
    // いずれの場合にもマッチしない場合、Errorを未対処のままにする
    throw e;
  }
}
```

### 類似するエラーと区別する

コードブロックは異なる対処が必要な理由で失敗するにも関わらず、非常によく似たエラー（つまり、同じ型やメッセージ）を投げることがしばしばあります。

あなたが発生した元々のエラーを管理していない場合、エラーを捕捉して、より詳細なメッセージを持つ新しい `Error` オブジェクトを発生させることが一つの選択肢となります。
元のエラーは新しい `Error` のコンストラクターの [`options`](/ja/docs/Web/JavaScript/Reference/Global_Objects/Error/Error#options) パラメーターの `cause` プロパティに渡すべきです。これによって、上位の try/catch ブロックが元のエラーとスタックトレースが利用できることを保証します。

> 以下の例は、似たエラーで失敗する2つのメソッドを示しています（`doFailSomeWay()` と `doFailAnotherWay()`）:

Reviewer: 数字の前後は空白を空けるようお願いいたします。
```suggestion
以下の例は、似たエラーで失敗する 2 つのメソッドを示しています（`doFailSomeWay()` と `doFailAnotherWay()`）:
```

    case "Failed in some way":
      handleFailSomeWay(err.cause);
      break;
    case "Failed in another way":
      handleFailAnotherWay(err.cause);
      break;
  }
}
```

> **メモ:** もしあなたがライブラリを制作している場合、利用者にエラーメッセージをパースするようお願いするかわりに、発生したエラーの微妙な差を区別するためにエラーの cause を使用すべきです。例については [Error の cause ページ](/ja/docs/Web/JavaScript/Reference/Global_Objects/Error/cause#providing_structured_data_as_the_error_cause) をご覧ください。

サブクラスのコンストラクターが `super()` を呼び出すときに `options` パラメーターを渡すと、[独自のエラー型](#独自のエラー型)も `cause` プロパティを利用できます。基底クラスのコンストラクター `Error()` は `options.cause` を読み取って、新しいエラーのインスタンスに `cause` プロパティを定義します。

```js
class MyError extends Error {
  constructor(message, options) {
    // "cause" プロパティを設定するために第二引数に `options` を渡す必要がある。
    super(message, options);
  }
}

console.log(new MyError("test", { cause: new Error("cause") }).cause);
// Error: cause
```

### 独自のエラー型

`Error` から派生した独自のエラー型を定義して `throw new CustomError()` ができるようにし、`instanceof CustomError` で例外ハンドラー内のエラーの種類を確認したいでしょう。これを行う一般的な方法の実例を以下に示します。

StackOverflow の突っ込んだ議論、 ["What's a good way to extend Error in JavaScript?"](http://stackoverflow.com/questions/1382107/whats-a-good-way-to-extend-error-in-javascript) も参照してください。

> > **警告**: 組み込みのサブクラス化は、ES6より古いコードに確実にトランスパイルできるわけではありません。なぜなら {{jsxref("Reflect.construct()")}} を使わずに特定の `new.target` を持つ基底クラスを構築する手段がないためです。[追加の設定](https://github.com/loganfsmyth/babel-plugin-transform-builtin-extend)を行うか、コンストラクタの最後で手動で {{jsxref("Object/setPrototypeOf", "Object.setPrototypeOf(this, CustomError.prototype)")}} を呼ぶ必要があります。そうしないと、構築されたインスタンスは `CustomError` のインスタンスになりません。詳しくは [the TypeScript FAQ](https://github.com/microsoft/TypeScript/wiki/FAQ#why-doesnt-extending-built-ins-like-error-array-and-map-work) をご覧ください。

Reviewer: こちらも英数字の前後に空白を入れるようお願いします。
また、-erで終わる単語には長音記号を付けてください。

```suggestion
> **警告**: 組み込みのサブクラス化は、ES6 より古いコードに確実にトランスパイルできるわけではありません。なぜなら {{jsxref("Reflect.construct()")}} を使わずに特定の `new.target` を持つ基底クラスを構築する手段がないためです。[追加の設定](https://github.com/loganfsmyth/babel-plugin-transform-builtin-extend)を行うか、コンストラクターの最後で手動で {{jsxref("Object/setPrototypeOf", "Object.setPrototypeOf(this, CustomError.prototype)")}} を呼ぶ必要があります。そうしないと、構築されたインスタンスは `CustomError` のインスタンスになりません。詳しくは [the TypeScript FAQ](https://github.com/microsoft/TypeScript/wiki/FAQ#why-doesnt-extending-built-ins-like-error-array-and-map-work) をご覧ください。
```

> > **メモ:** ES2015 クラスを使用した場合、一部のブラウザはスタックトレース上に `CustomError` コンストラクタを含めます。

Reviewer: ```suggestion
> **メモ:** ES2015 クラスを使用した場合、一部のブラウザはスタックトレース上に `CustomError` コンストラクターを含めます。
```

```js
class CustomError extends Error {
  constructor(foo = "bar", ...params) {

>     // 親のコンストラクタに（ベンダー固有のものも含めて）残りの引数を渡す

Reviewer: ```suggestion
    // 親のコンストラクターに（ベンダー固有のものも含めて）残りの引数を渡す
```
